### PR TITLE
L015: Handle COUNT(DISTINCT(col))

### LINKED ISSUE #3823: L015: DISTINCT with parenthesis is not found when inside a COUNT fn
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

Rule L015 (which should flag usages of `DISTINCT` keyword that looks like a function call), does not seem to find instances where it is violated, where the `DISTINCT` keyword is inside an aggregation function.

### Expected Behaviour

Expected these non-conforming `DISTINCT` usages to be flagged.

### Observed Behaviour

Linter is happy, does not report any errors.

### How to reproduce

Copy-paste sample into a file, e.g. `demo.sql`

```
SELECT count(distinct(unique_key)) AS total_count
FROM some_table
GROUP BY unique_key;
```

Run
```
sqlfluff lint --rules L015 --dialect ansi demo.sql
```

### Dialect

Tried this on Ansi and also Sparksql, but I don't think it matters.

### Version

sqlfluff, version 1.3.0

Python 3.9.13

### Configuration

not needed -- only tested one rule, as specified on command line arg

### Are you willing to work on and submit a PR to address the issue?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #3823

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by barrywhart** [comment]
@alanmcruickshank: When you have a bit of time, could you look at the failing test case `test/rules/yaml_test_cases_test.py::test__rule_test_case[L015_test_fail_distinct_concat_inside_count]`? Reflow is behaving oddly.

Input query:
```
SELECT COUNT(DISTINCT(CONCAT(col1, '-', col2, '-', col3)))
```

Expected fix:
```
SELECT COUNT(DISTINCT CONCAT(col1, '-', col2, '-', col3))
```

Actual fix (note the space after `CONCAT`):
```
SELECT COUNT(DISTINCT CONCAT (col1, '-', col2, '-', col3))
```

If you run `sqlfluff fix` from the command line, you'll see that the first run ends with:
```
[1 unfixable linting violations found]
```

Running `sqlfluff fix` a second time fixes the "unfixable" violation and gives the expected output.

File: src/sqlfluff/rules/L015.py [hunk 1]
```diff
@@ -35,46 +36,81 @@ class Rule_L015(BaseRule):
...
-            else:
-                anchor = modifier[0]
-                seq = ReflowSequence.from_around_target(
-                    modifier[0],
-                    context.parent_stack[0],
-                    config=context.config,
-                    sides="after",
-                )
+        if context.segment.is_type("select_clause"):
+            # Look for `select_clause_modifier`
```
Comment (barrywhart) [inline src/sqlfluff/rules/L015.py:50]: This section is the original case (prior to this PR).

File: src/sqlfluff/rules/L015.py [hunk 2]
```diff
@@ -35,46 +36,81 @@ class Rule_L015(BaseRule):
...
+                else:
+                    anchor = modifier[0]
+                    seq = ReflowSequence.from_around_target(
+                        modifier[0],
+                        context.parent_stack[0],
+                        config=context.config,
+                        sides="after",
+                    )
+        elif context.segment.is_type("function"):
+            # Look for a function call DISTINCT() whose parent is an expression
```
Comment (barrywhart) [inline src/sqlfluff/rules/L015.py:73]: This section is the new case.

File: src/sqlfluff/rules/L015.py [hunk 3]
```diff
@@ -35,46 +36,81 @@ class Rule_L015(BaseRule):
...
+                return None
+            function_name = children.select(sp.is_type("function_name")).first()
+            bracketed = children.first(sp.is_type("bracketed"))
+            if (
+                not function_name
+                or function_name[0].raw_upper != "DISTINCT"
+                or not bracketed
+            ):
+                return None
+            anchor = context.parent_stack[-1]
```
Comment (greg-finley) [inline src/sqlfluff/rules/L015.py:88]: Better to define `anchor` at line 75 and then check `anchor` in `if` on lines 76 and 77?

Comment (barrywhart) [reply src/sqlfluff/rules/L015.py:88]: Good point -- I'll change it.

**Comment by barrywhart** [comment]
@greg-finley: Ready for another review (assuming the build passes).

**Review by greg-finley — APPROVED** [review]

**Pull request merged**